### PR TITLE
for fixing potential TOCTOU issue

### PR DESCRIPTION
**Pull request by danielwangksu**
fx https://github.com/ros/rospack/issues/103

`lstat` does not follow symbolic link but `fstat` does, comparing mode make sure there is no TOCTOU race condition

### CODE REVIEW

File: src/rospack.cpp [hunk 1]
```diff
@@ -2112,20 +2112,30 @@ Rosstackage::validateCache()
...
+    return NULL;
+
   // try to open it
   FILE* cache = fopen(cache_path.c_str(), "r");
   if(!cache)
     return NULL; // it's not readable by us. sad.
 
+  struct stat s;
+  if(fstat(fileno(cache), &s) == -1)
+    return NULL;
```
Comment (dirk-thomas) [inline src/rospack.cpp:2134]: @danielwangksu Doesn't this need a `fclose(cache)` here?

**Review by dirk-thomas — COMMENTED** [review]
Thank you for creating the pull request.

I added a few commits to (fix indentation, avoid unnecessary changes to the logic / indentation level / whitespaces) to make the diff much smaller are more readable.

Just one question inline...

File: src/rospack.cpp (hunk 1 above)
Comment (danielwangksu) [reply src/rospack.cpp:2134]: Yes you are right. Sorry missed this.

Comment (danielwangksu) [reply src/rospack.cpp:2134]: Thank you very much for helping fix this :)

**Comment by dirk-thomas** [comment]
Thanks for the patch and the quick response!

**Pull request merged**